src/content/drafts/sourcecode-to-tokens-AST.md

```diff
@@ -0,0 +1,97 @@
+---
+layout: post
+title: How a parser translates text into tokens and then into an abstract syntax tree
+teaser: "You will learn how the source code in any proramming language is converted to multiple tokens and is then parsed to generate an Abstract Syntax Tree. You can then then traverse the AST to write and build custom rules in ESLint."
+image: 
+
+authors:
+  - gayathri-krishnaswamy
+categories:
+  - Tutorials
+tags:
+  - Lexical analysis
+  - Syntax analysis
+  - Abstract Syntax Tree
+  - AST
+  - Tokens
+  - ESLint rules
+---
+##  Introduction
+A parser is a program that takes input in the form of sequential source program instructions and breaks it into smaller pieces (tokens) to generate statements and expressions. The parser then represents the statements and expressions in the form of a parse tree or an Abstract Syntax Tree (AST) that depicts the syntactic structure of your source code.
+
+:information_source: Not all parsers create an AST. An AST is required only to inspect the overall structure of the program. Parsers written for other purposes, such as immediate conversion into another format, may not create an AST at all.
+
+In this tutorial, first, you'll learn on a high-level how your source code is broken down into multiple tokens and is then converted to an AST. Next, you'll learn how [ESLint](https://eslint.org/docs/latest/) leverages the capability of AST and helps you build your own custom rules. This tutorial is intended for those who use ESLint and wants to understand more about how the tool works. It assumes that you have basic familiarity with:
+- Computer programming
+- Javascript coding
+- ESLint tool
+  
+By the end of this tutorial, you’ll be able to:
+- Understand how the source code is converted to AST.
+- Generate AST for your source program using AST explorer.
+- Understand how to traverse the AST nodes to write custom rules in ESLint.
+
+## Step 1: Lexical analysis
+[Lexical analysis](https://en.wikipedia.org/wiki/Lexical_analysis) commonly known as tokenization is the process of converting your source code into a sequence of tokens. A token can be any valid entity in a programming language such as an identifier, keyword, operator, function call, string, or separator. Tokens do not understand how things fit together instead focus on different components of a file. Consider it analogous to breaking a sentence into different words and identifying the words such as nouns, verbs, numbers, adjectives, and punctuation. At this point, you don’t understand what’s part of a sentence or how these words fit together to make a sentence. The below picture depicts how a simple `if statement` is converted into multiple tokens.
+
+![If statement converted to tokens](/src/assets/images/blog/2022/lexical-analysis.png "source code to tokens")
```

Comment:
https://github.com/eslint/eslint.org/pull/324/files#r954747406
```suggestion
![If statement converted to tokens](/assets/images/blog/2022/lexical-analysis.png "source code to tokens")
```

Reply:
Changed the path. Can you please check in preview mode?